src/util/string.h

```diff
@@ -17,6 +17,28 @@
 #include <vector>
 
 namespace util {
+/** Type to denote a format string that was checked at compile time */
+template <int num_params>
+struct ConstevalFormatString {
+    const char* const fmt;
+    consteval ConstevalFormatString(const char* str) : fmt{str} { CheckNumFormatSpecifiers(fmt); }
+    consteval static void CheckNumFormatSpecifiers(std::string_view str)
+    {
+        int count{0};
+        bool begin_f{false};
+        for (char c : str) {
+            bool perc{c == '%'};
+            if (begin_f) {
+                count += !perc;
+                begin_f = false;
+                continue;
+            }
+            if (perc) begin_f = true;
+        }
+        assert(num_params == count);
```

Comment:
I think it would be good practice to add a comment directly behind these asserts in consteval functions explaining what the error is, so we get better error messages e.g.
```
assert(num_params == count); // Format arguments have to match the number of format specifiers in the string.
```